File: handler/converter.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-
"""
    将Excel中的数据转化为rpy中的对象
"""
from collections import namedtuple

from const.converter_setting import ElementColNumMapping, PositionMapping, ImageCmdMapping, TransitionMapping, \
    ReplaceCharacterMapping
from model.element import Text, Image, Transition, Audio, Role, Command, Voice, Menu

SheetConvertResult = namedtuple('SheetConvertResult', ['label', 'data'])

RowConvertResult = namedtuple('RowConvertResult',
                              ['role',  # 角色
                               'mode',  # 模式
                               'text',  # 文本
                               'music',  # 音乐
                               'character',  # 立绘
                               'change_page',  # 换页
                               'background',  # 背景
                               'remark',  # 备注
                               'sound',  # 音效
                               'transition',  # 转场
                               'voice',  # 语音
                               'menu',  # 条件跳转
                               'side_character'  # 头像
                               ])


class Converter(object):

    def __init__(self, parser):
        self.parser = parser
        self.roles = list()
        self.role_name_mapping = dict()
        self.current_mode = 'nvl'
        self.current_role = Role("narrator_nvl", "None")
        self.characters = list()
        self.side_characters = dict()

    def add_role(self, name):
        role = self.role_name_mapping.get(name)
        if not role:
            role = Role("role{}".format(len(self.role_name_mapping.keys()) + 1), name)
            self.role_name_mapping[name] = role
        return role

    #创建一个元组，存有工作表标签及对应工作表下的多行转换后数据
    def generate_rpy_elements(self):
        result = []
        parsed_sheets = self.parser.get_parsed_sheets()
        for idx, parsed_sheet in enumerate(parsed_sheets):
            if idx == 0:
                label = 'start'
            else:
                label = parsed_sheet.name
            result.append(SheetConvertResult(label=label, data=self.parse_by_sheet(parsed_sheet.row_values, idx)))
        return result

    @classmethod
    def generate_character(cls, img_str):
        last_word = img_str.split(" ")[-1]
        position = PositionMapping.get(last_word, None) or last_word
        if position:
            return Image(img_str.replace(last_word, "").strip(), "show", position)
        else:
            return Image(img_str.replace(last_word, "").strip(), ImageCmdMapping.get(last_word, "hide"))

    #循环调用parse_by_row_value方法，返回拼接多行转换后信息的列表
    def parse_by_sheet(self, values, sheet_index):
        result = []
        current_role_name = None  # 用于跟踪最近的有效 role_name
        for row_index, row_value in enumerate(values):
            role_name = row_value[ElementColNumMapping.get('role_name')]
            if role_name.strip():
                current_role_name = role_name  # 更新最近的有效 role_name
            else:
                role_name = current_role_name  # 如果当前 role_name 为空，使用最近的有效值
            result.append(self.parse_by_row_value(row_value, role_name, sheet_index, row_index))
        return result

    #调用RowConverter的convert方法，返回存有单行转换后信息的元组
    def parse_by_row_value(self, row, role_name, sheet_index, row_index):
        row_converter = RowConverter(row, self, role_name, sheet_index, row_index)
        return row_converter.convert()


class RowConverter(object):

    def __init__(self, row, converter, role_name, sheet_index, row_index):
        self.row = row
        self.converter = converter
        self.role_name = role_name  
        self.row_index = row_index  
        self.sheet_index = sheet_index
    
    #该方法返回存有单行转换后信息的元组
    def convert(self):
        return RowConvertResult(
            mode=self._converter_mode(),
            role=self._converter_role(),
            text=self._converter_text(),
            music=self._converter_music(),
            character=self._converter_character(),
            change_page=self._converter_change_page(),
            background=self._converter_background(),
            remark=self._converter_remark(),
            sound=self._converter_sound(),
            transition=self._converter_transition(),
            voice=self._converter_voice(),
            menu=self._converter_menu(),
            side_character=self._converter_side_character(),
        )

    def _converter_mode(self):
        # 模式
        mode = self.row[ElementColNumMapping.get('mode')]
        if mode:
            self.converter.current_mode = mode
        return mode

    def _converter_role(self):
        # 角色
        role_name = self.row[ElementColNumMapping.get('role_name')]
        if role_name and role_name != "旁白":
            # 当新的角色名出现时，切换到该角色
            self.converter.current_role = self.converter.add_role(role_name)
            #self.converter.current_mode = "nvl"  # 可选：根据需要设置当前模式
        elif role_name == "":
            # 空角色名时，保持当前角色不变
            return self.converter.current_role
        else:
            # 处理旁白角色或其他情况
            self.converter.current_role = Role("narrator_{}".format(self.converter.current_mode), "None")
        
        return self.converter.current_role

    def _converter_text(self):
        # 文本
        text = str(self.row[ElementColNumMapping.get('text')]).replace("\n", "\\n")
        if not text:
            return None
        replace_index_char = []
        for idx, t in enumerate(text):
            if ReplaceCharacterMapping.get(t):
                replace_index_char.append((idx, t))

        if replace_index_char:
            new_text_list = list(text)
            for idx, char in replace_index_char:
                new_text_list[idx] = ReplaceCharacterMapping.get(char)
            text = ''.join(new_text_list)
        return Text(text, self.converter.current_role)

    def _converter_music(self):
        # 音乐
        music = self.row[ElementColNumMapping.get('music')]
        if not music:
            return None
        cmd = "stop" if music == "none" else "play"
        return Audio(music, cmd)

    def _converter_background(self):
        # 背景
        background = self.row[ElementColNumMapping.get('background')]
        if not background:
            return None
        return Image(background, "scene")

    def _converter_character(self):
        character_str = str(self.row[ElementColNumMapping['character']]).strip()
        # --- 1. 统一地回收旧立绘 ---
        hide_images = [Image(char.name, 'hide') for char in self.converter.characters]
        # --- 2. 若本行没有立绘，只需回收后结束 ---
        if not character_str:
            # 清空缓存，避免残留
            self.converter.characters = []
            return hide_images            # 只返回“hide”指令
        # --- 3. 解析并生成新立绘 ---
        new_characters = [
            Converter.generate_character(ch)
            for ch in character_str.split(';') if ch.strip()
        ]
        # 更新缓存为当前行的新立绘
        self.converter.characters = new_characters
        # 返回：先隐藏旧立绘，再展示新立绘
        return hide_images + new_characters

    def _converter_remark(self):
        pass

    def _converter_sound(self):
        # 音效
        sound = self.row[ElementColNumMapping.get('sound')]
        if not sound:
            return None
        if sound.startswith('循环'):
            return Audio(sound.replace('循环', ''), 'loop')
        else:
            cmd = "stop" if sound == "stop" else "sound"
            return Audio(sound, cmd)

    def _converter_transition(self):
        # 转场
        transition = self.row[ElementColNumMapping.get('transition')]
        if not transition:
            return None
        t_style = TransitionMapping.get(transition, "")
        return Transition(t_style)

    def _converter_change_page(self):
        # 换页
        change_page = self.row[ElementColNumMapping.get('change_page')]
        if not change_page:
            return None
        return Command("nvl clear")

    def _converter_voice(self):
        voice_str = str(self.row[ElementColNumMapping.get('voice')]).strip()
        if not voice_str:
            return None
        
        # 检查是否包含 "tts"
        if voice_str.lower().strip() == "tts":
            return Voice(f"{self.role_name}_sheet{self.sheet_index+1}_row{self.row_index+8}_synthesized.wav")
        
        if voice_str.split(" ")[-1] == "sustain":
            voice_name = voice_str.split(" ")[0]
            return Voice(voice_name, sustain=True)
        else:
            return Voice(voice_str)

    def _converter_menu(self):
        # 分支条件的label写在对话文本列
        menu = self.row[ElementColNumMapping.get('menu')]
        if not menu:
            return None
        text = str(self.row[ElementColNumMapping.get('text')]).replace("\n", "\\n")
        if not text:
            return None
        replace_index_char = []
        for idx, t in enumerate(text):
            if ReplaceCharacterMapping.get(t):
                replace_index_char.append((idx, t))

        if replace_index_char:
            new_text_list = list(text)
            for idx, char in replace_index_char:
                new_text_list[idx] = ReplaceCharacterMapping.get(char)
            text = ''.join(new_text_list)
        return Menu(label=text, target=menu)

    def _converter_side_character(self):
        # 对话框头像
        character_str = str(self.row[ElementColNumMapping.get('side_character')]).strip()
        if not character_str:
            return None
        self.converter.side_characters[self.converter.current_role.pronoun] = character_str
        return None

    def _converter_voice_cmd(self):
        pass

File: packages/core/src/convert/converter.ts
// 行→元素转换（移植自 handler/converter.py）。逐字符对齐旧行为，见 docs/01-legacy-system-contract.md §3。
import {
  ElementColNumMapping,
  PositionMapping,
  ImageCmdMapping,
  TransitionMapping,
  ReplaceCharacterMapping,
  type ColKey,
} from '../settings/converterSetting'
import { asStr, truthy, rawEq, EMPTY, type CellValue } from '../parse/cellValue'
import { Role, Text, Image, Transition, Audio, Voice, Menu, Command } from '../model/element'

export interface ParsedSheet {
  name: string
  rows: CellValue[][]
}

export interface RowConvertResult {
  role: Role
  modeRaw: string // 旧 RowConvertResult.mode，下游未使用
  text: Text | null
  music: Audio | null
  character: Image[] | null
  changePage: Command | null
  background: Image | null
  remark: null
  sound: Audio | null
  transition: Transition | null
  voice: Voice | null
  menu: Menu | null
  sideCharacter: null
}

export interface SheetConvertResult {
  label: string
  data: RowConvertResult[]
}

export type ConversionMode = 'legacy-compat' | 'default'

export interface PipelineOptions {
  mode: ConversionMode
  // default 模式下的修正项（legacy-compat 下应为 false 以保持与黄金样本逐字符一致）
  normalizeMode?: boolean // 修正 bug：模式大小写（ADV→adv）
  trimRoleNames?: boolean // 修正 bug：角色名首尾空白（朝比奈实玖瑠\t）
}

export interface ConvertWarning {
  code: string
  message: string
  sheetIndex: number
  rowIndex: number // 0 基解析行序号；Excel 行号 = rowIndex + 8
}

function col(row: CellValue[], key: ColKey): CellValue {
  return row[ElementColNumMapping[key]] ?? EMPTY
}

// 单次逐字符转义（不可链式 replace，否则二次转义）。
function escapeText(s: string): string {
  let out = ''
  for (const ch of s) out += ReplaceCharacterMapping[ch] ?? ch
  return out
}

export class Converter {
  roleNameMapping = new Map<string, Role>() // 角色名 → Role，按首次出现顺序（跨 sheet 累积）
  currentMode = 'nvl'
  currentRole: Role = new Role('narrator_nvl', 'None')
  characters: Image[] = [] // 当前在场立绘缓存
  sideCharacters = new Map<string, string>() // pronoun → 头像路径
  warnings: ConvertWarning[] = []

  constructor(
    public sheets: ParsedSheet[],
    public opts: PipelineOptions,
  ) {}

  addRole(name: string): Role {
    let role = this.roleNameMapping.get(name)
    if (!role) {
      role = new Role(`role${this.roleNameMapping.size + 1}`, name)
      this.roleNameMapping.set(name, role)
    }
    return role
  }

  warn(code: string, message: string, sheetIndex: number, rowIndex: number): void {
    this.warnings.push({ code, message, sheetIndex, rowIndex })
  }

  generate(): SheetConvertResult[] {
    const result: SheetConvertResult[] = []
    this.sheets.forEach((sheet, idx) => {
      const label = idx === 0 ? 'start' : sheet.name
      result.push({ label, data: this.parseSheet(sheet.rows, idx) })
    })
    return result
  }

  private parseSheet(rows: CellValue[][], sheetIndex: number): RowConvertResult[] {
    const out: RowConvertResult[] = []
    // 角色名前向填充：仅用于 TTS 语音文件名（_converter_role 读原始单元格，不读这个值）。
    let currentRoleName: string | null = null
    rows.forEach((row, rowIndex) => {
      const roleCell = col(row, 'role_name')
      let roleName: string | null
      if (asStr(roleCell).trim()) {
        roleName = asStr(roleCell)
        currentRoleName = roleName
      } else {
        roleName = currentRoleName
      }
      out.push(new RowConverter(this, row, roleName, sheetIndex, rowIndex).convert())
    })
    return out
  }
}

class RowConverter {
  constructor(
    private conv: Converter,
    private row: CellValue[],
    private roleName: string | null, // 前向填充值，仅 voice tts 用
    private sheetIndex: number,
    private rowIndex: number,
  ) {}

  private cell(key: ColKey): CellValue {
    return col(this.row, key)
  }

  // 字段求值顺序必须与旧 namedtuple 构造顺序一致（含副作用：mode→role→...）。
  convert(): RowConvertResult {
    const modeRaw = this.convMode()
    const role = this.convRole()
    const text = this.convText()
    const music = this.convMusic()
    const character = this.convCharacter()
    const changePage = this.convChangePage()
    const background = this.convBackground()
    const remark = null
    const sound = this.convSound()
    const transition = this.convTransition()
    const voice = this.convVoice()
    const menu = this.convMenu()
    const sideCharacter = this.convSideCharacter()
    return {
      role,
      modeRaw,
      text,
      music,
      character,
      changePage,
      background,
      remark,
      sound,
      transition,
      voice,
      menu,
      sideCharacter,
    }
  }

  private convMode(): string {
    const m = this.cell('mode')
    if (truthy(m)) {
      let v = asStr(m)
      if (this.conv.opts.normalizeMode) {
        const lower = v.toLowerCase()
        if ((lower === 'nvl' || lower === 'adv') && lower !== v) {
          this.conv.warn(
            'mode-case',
            `模式 "${v}" 已归一化为 "${lower}"`,
            this.sheetIndex,
            this.rowIndex,
          )
          v = lower
        }
      }
      this.conv.currentMode = v
    }
    return asStr(m)
  }

  private convRole(): Role {
    const raw = this.cell('role_name')
    const rawStr = asStr(raw)
    if (truthy(raw) && rawStr !== '旁白') {
      let name = rawStr
      if (this.conv.opts.trimRoleNames) {
        const t = name.trim()
        if (t !== name) {
          this.conv.warn(
            'role-name-whitespace',
            `角色名含首尾空白，已修剪为 "${t}"`,
            this.sheetIndex,
            this.rowIndex,
          )
          name = t
        }
      }
      this.conv.currentRole = this.conv.addRole(name)
    } else if (rawStr === '') {
      // 空角色名：保持 currentRole 不变
    } else {
      this.conv.currentRole = new Role(`narrator_${this.conv.currentMode}`, 'None')
    }
    return this.conv.currentRole
  }

  private convText(): Text | null {
    const t0 = asStr(this.cell('text')).replace(/\n/g, '\\n')
    if (!t0) return null
    return new Text(escapeText(t0), this.conv.currentRole)
  }

  private convMusic(): Audio | null {
    const m = this.cell('music')
    if (!truthy(m)) return null
    const cmd = rawEq(m, 'none') ? 'stop' : 'play'
    return new Audio(m, cmd)
  }

  private convBackground(): Image | null {
    const b = this.cell('background')
    if (!truthy(b)) return null
    return new Image(asStr(b), 'scene')
  }

  private convCharacter(): Image[] {
    const cStr = asStr(this.cell('character')).trim()
    // 1. 统一回收旧立绘
    const hides = this.conv.characters.map((ch) => new Image(ch.name, 'hide'))
    // 2. 本行无立绘：清空缓存，仅回收
    if (!cStr) {
      this.conv.characters = []
      return hides
    }
    // 3. 解析新立绘（按原始书写顺序）
    const news = cStr
      .split(';')
      .filter((seg) => seg.trim())
      .map((seg) => RowConverter.generateCharacter(seg))
    this.conv.characters = news
    return [...hides, ...news]
  }

  static generateCharacter(imgStr: string): Image {
    const tokens = imgStr.split(' ')
    const lastWord = tokens[tokens.length - 1] ?? ''
    const position = PositionMapping[lastWord] ?? lastWord
    // 旧代码用 Python str.replace（替换全部出现），故用 replaceAll。
    const name = imgStr.replaceAll(lastWord, '').trim()
    if (position) {
      return new Image(name, 'show', position)
    }
    return new Image(name, ImageCmdMapping[lastWord] ?? 'hide')
  }

  private convSound(): Audio | null {
    const s = this.cell('sound')
    if (!truthy(s)) return null
    const str = asStr(s)
    if (str.startsWith('循环')) {
      return new Audio(str.replaceAll('循环', ''), 'loop')
    }
    const cmd = rawEq(s, 'stop') ? 'stop' : 'sound'
    return new Audio(s, cmd)
  }

  private convTransition(): Transition | null {
    const t = this.cell('transition')
    if (!truthy(t)) return null
    const style = TransitionMapping[asStr(t)] ?? ''
    return new Transition(style)
  }

  private convChangePage(): Command | null {
    const c = this.cell('change_page')
    if (!truthy(c)) return null
    return new Command('nvl clear')
  }

  private convVoice(): Voice | null {
    const vStr = asStr(this.cell('voice')).trim()
    if (!vStr) return null
    if (vStr.toLowerCase() === 'tts') {
      const rn = this.roleName === null ? 'None' : this.roleName
      return new Voice(`${rn}_sheet${this.sheetIndex + 1}_row${this.rowIndex + 8}_synthesized.wav`)
    }
    const tokens = vStr.split(' ')
    if (tokens[tokens.length - 1] === 'sustain') {
      return new Voice(tokens[0] ?? '', true)
    }
    return new Voice(vStr)
  }

  private convMenu(): Menu | null {
    const m = this.cell('menu')
    if (!truthy(m)) return null
    const t0 = asStr(this.cell('text')).replace(/\n/g, '\\n')
    if (!t0) return null
    return new Menu(escapeText(t0), asStr(m))
  }

  private convSideCharacter(): null {
    const s = asStr(this.cell('side_character')).trim()
    if (!s) return null
    this.conv.sideCharacters.set(this.conv.currentRole.pronoun, s)
    return null
  }
}
